### PR TITLE
ARROW-17980: [C++] As-of-Join Substrait extension

### PR DESCRIPTION
**Pull request by rtpsw**
See https://issues.apache.org/jira/browse/ARROW-17980

### CODE REVIEW

**Comment by rtpsw** [comment]
cc @icexelloss @westonpace

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 1]
```diff
@@ -3187,5 +3198,164 @@ TEST(Substrait, IsthmusPlan) {
...
+                    }
+                  },
+                  "rootReference": {}
+                }
+              },
+              "by": [
+                {
+                  "selection": {
+                    "directReference": {
+                      "structField": {
```
Comment (icexelloss) [inline cpp/src/arrow/engine/substrait/serde_test.cc:3309]: What does the "structField" mean here?

Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/serde_test.cc:3309]: It means access to a field by index. It leads Arrow Substrait to create a `FieldRef` with the field index given inside.

**Comment by rtpsw** [comment]
Note that this PR includes code from https://github.com/apache/arrow/pull/14386 which is currently pending.

**Comment by icexelloss** [comment]
@rtpsw Why do we need #14386 for this? Can we separate out these two changes?

**Comment by icexelloss** [comment]
@westonpace The changes here depends on your local change in https://github.com/westonpace/arrow/tree/experiment/substrait-extension

How do you want to proceed here? We can either 
(1) merge changes in experiment/substrait-extension and rebase this PR
(2) combine changes in experiment/substrait-extension and the asof join message together in this PR and review together

**Comment by rtpsw** [comment]
> @rtpsw Why do we need #14386 for this? Can we separate out these two changes?

Technically, we don't, or no longer do. I used #14386 during test cases development of this PR. If #14386 will be rejected, we could remove its code here too. Otherwise, the code can stay here and will merge cleanly with #14386.

**Comment by rtpsw** [comment]
@westonpace, any idea why some jobs, like [this one](https://github.com/apache/arrow/actions/runs/3240443758/jobs/5311073777), failed? It seems to be complaining about missing proto classes; I don't know why the proto classes would get built in some jobs and not in others.

File: cpp/proto/substrait/extension_rels.proto [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+import "substrait/algebra.proto";
+
+option csharp_namespace = "Arrow.Substrait";
+option go_package = "github.com/apache/arrow/substrait";
+option java_multiple_files = true;
+option java_package = "io.arrow.substrait";
+
+message AsOfJoinRel {
+  .substrait.Expression on = 1;
```
Comment (icexelloss) [inline cpp/proto/substrait/extension_rels.proto:29]: This requires all the `on` column must have the same index for all tables?

File: cpp/src/arrow/compare.cc [hunk 3]
```diff
@@ -305,6 +305,11 @@ class RangeDataEqualsImpl {
   Status Visit(const StructType& type) {
     const int32_t num_fields = type.num_fields();
 
+    if (left_.child_data.size() != static_cast<size_t>(num_fields) ||
```
Comment (icexelloss) [inline cpp/src/arrow/compare.cc:308]: Can we please separate out orthogonal changes?

File: cpp/src/arrow/compute/exec/asof_join_node.cc [hunk 4]
```diff
@@ -951,28 +953,27 @@ class AsofJoinNode : public ExecNode {
   }
 
   static arrow::Result<std::shared_ptr<Schema>> MakeOutputSchema(
-      const std::vector<ExecNode*>& inputs,
+      const std::vector<std::shared_ptr<Schema>> input_schema,
```
Comment (icexelloss) [inline cpp/src/arrow/compute/exec/asof_join_node.cc:956]: Why this change?

File: cpp/src/arrow/compute/exec/asof_join_node.cc [hunk 5]
```diff
@@ -1030,6 +1031,32 @@ class AsofJoinNode : public ExecNode {
     return match.indices()[0];
   }
 
+  static Result<std::vector<col_index_t>> GetIndicesOfOnKey(
+      const std::vector<std::shared_ptr<Schema>>& input_schema, const FieldRef& on_key) {
+    size_t n_input = input_schema.size();
+    std::vector<col_index_t> indices_of_on_key(n_input);
+    for (size_t i = 0; i < n_input; ++i) {
+      ARROW_ASSIGN_OR_RAISE(indices_of_on_key[i],
+                            FindColIndex(*input_schema[i], on_key, "on"));
```
Comment (icexelloss) [inline cpp/src/arrow/compute/exec/asof_join_node.cc:1040]: I thought substrait/FieldRef is index based already - why do we need to find index here?

File: cpp/proto/substrait/extension_rels.proto (hunk 2 above)
Comment (rtpsw) [reply cpp/proto/substrait/extension_rels.proto:29]: The proto requires a common expression, which could be a name rather than an index. OTOH, the Arrow Substrait code in this PR only deals with a field index expression; this can be extended, if desired.

File: cpp/src/arrow/compare.cc (hunk 3 above)
Comment (rtpsw) [reply cpp/src/arrow/compare.cc:308]: Will do.

File: cpp/src/arrow/compute/exec/asof_join_node.cc (hunk 4 above)
Comment (rtpsw) [reply cpp/src/arrow/compute/exec/asof_join_node.cc:956]: This allows exposing `MakeOutputSchema` to a caller that has a schema rather than an `ExecNode`, as is done in the tester.

File: cpp/src/arrow/compute/exec/asof_join_node.cc (hunk 5 above)
Comment (rtpsw) [reply cpp/src/arrow/compute/exec/asof_join_node.cc:1040]: The proto uses `Expression` to allow for various ways to specify the key; it can be thought of as future-proof. We can choose which of these ways to support in Arrow Substrait in a first version. The code here supports any kind of `FieldRef`, be it named or indexed, that Arrow Substrait may set up. The Arrow Substrait code only supports an indexed `FieldRef` (see [here](https://github.com/apache/arrow/blob/e8d54ea765ae7ba63b8f42c29ec855d656e85dc8/cpp/src/arrow/engine/substrait/expression_internal.cc#L150)), but it can be extended.

File: cpp/proto/substrait/extension_rels.proto (hunk 2 above)
Comment (icexelloss) [reply cpp/proto/substrait/extension_rels.proto:29]: We are using `StructField` for this so I don't think we can use name for this, right?

I think it probably makes more sense to have this only be a repeated field in case that `on` column doesn't share the same column index for each table

File: cpp/src/arrow/compute/exec/asof_join_node.cc (hunk 4 above)
Comment (icexelloss) [reply cpp/src/arrow/compute/exec/asof_join_node.cc:956]: kk sounds reasonable

File: cpp/src/arrow/compute/exec/asof_join_node.cc [hunk 6]
```diff
@@ -951,28 +953,27 @@ class AsofJoinNode : public ExecNode {
...
 
   static arrow::Result<std::shared_ptr<Schema>> MakeOutputSchema(
-      const std::vector<ExecNode*>& inputs,
+      const std::vector<std::shared_ptr<Schema>> input_schema,
       const std::vector<col_index_t>& indices_of_on_key,
       const std::vector<std::vector<col_index_t>>& indices_of_by_key) {
     std::vector<std::shared_ptr<arrow::Field>> fields;
 
-    size_t n_by = indices_of_by_key[0].size();
+    size_t n_by = indices_of_by_key.size() == 0 ? 0 : indices_of_by_key[0].size();
```
Comment (icexelloss) [inline cpp/src/arrow/compute/exec/asof_join_node.cc:961]: When can `indices_of_by_key.size() == 0` means this is a join without `by` key?

File: cpp/src/arrow/compute/exec/asof_join_node.cc (hunk 5 above)
Comment (icexelloss) [reply cpp/src/arrow/compute/exec/asof_join_node.cc:1040]: I am not sure we need this function - in practice this substrait field ref should always be index based..

File: cpp/src/arrow/compute/exec/asof_join_node.cc [hunk 7]
```diff
@@ -1030,6 +1031,32 @@ class AsofJoinNode : public ExecNode {
...
+    size_t n_input = input_schema.size();
+    std::vector<col_index_t> indices_of_on_key(n_input);
+    for (size_t i = 0; i < n_input; ++i) {
+      ARROW_ASSIGN_OR_RAISE(indices_of_on_key[i],
+                            FindColIndex(*input_schema[i], on_key, "on"));
+    }
+    return indices_of_on_key;
+  }
+
+  static Result<std::vector<std::vector<col_index_t>>> GetIndicesOfByKey(
```
Comment (icexelloss) [inline cpp/src/arrow/compute/exec/asof_join_node.cc:1045]: Same comment as `GetIndicesOfByKey`

File: cpp/src/arrow/engine/substrait/options.cc [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+      return Status::Invalid("substrait::AsOfJoinNode missing on-key");
+    }
+    ARROW_ASSIGN_OR_RAISE(auto on_key_expr, FromProto(as_of_join_rel.on(), ext_set, {}));
+    if (on_key_expr.field_ref() == NULLPTR) {
+      return Status::NotImplemented("substrait::AsOfJoinNode non-field-ref on-key");
+    }
+    const FieldRef& on_key = *on_key_expr.field_ref();
+
+    // by-key
+    std::vector<FieldRef> by_key;
```
Comment (icexelloss) [inline cpp/src/arrow/engine/substrait/options.cc:64]: I think `by_keys` here is a little clear since this is a vector

File: cpp/src/arrow/engine/substrait/options.h [hunk 9]
```diff
@@ -32,7 +36,7 @@ namespace engine {
 /// How strictly to adhere to the input structure when converting between Substrait and
 /// Acero representations of a plan. This allows the user to trade conversion accuracy
 /// for performance and lenience.
-enum class ConversionStrictness {
+enum class ARROW_ENGINE_EXPORT ConversionStrictness {
```
Comment (icexelloss) [inline cpp/src/arrow/engine/substrait/options.h:39]: What does the macro `ARROW_ENGINE_EXPORT` do?

File: cpp/src/arrow/engine/substrait/options.cc [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    for (size_t i = 0; i < inputs.size(); i++) {
+      input_decls[i] = inputs[i].declaration;
+    }
+    return DeclarationInfo{
+        compute::Declaration("asofjoin", input_decls, std::move(asofjoin_node_opts)),
+        std::move(schema)};
+  }
+};
+
+std::shared_ptr<ExtensionProvider> ExtensionProvider::kDefaultExtensionProvider =
```
Comment (icexelloss) [inline cpp/src/arrow/engine/substrait/options.cc:95]: Should we move this to substrait/options.h file? (Since the `kDefaultNamedTableProvider` is also there)

File: cpp/src/arrow/engine/substrait/options.h [hunk 11]
```diff
@@ -65,16 +69,27 @@ using NamedTableProvider =
     std::function<Result<compute::Declaration>(const std::vector<std::string>&)>;
 static NamedTableProvider kDefaultNamedTableProvider;
 
+class ARROW_ENGINE_EXPORT ExtensionProvider {
+ public:
+  static std::shared_ptr<ExtensionProvider> kDefaultExtensionProvider;
```
Comment (icexelloss) [inline cpp/src/arrow/engine/substrait/options.h:74]: Should we define this the same way that `kDefaultNamedTableProvider` is defined? I don't see a reason why those two should be different

**Review by icexelloss — CHANGES_REQUESTED** [review]

**Comment by icexelloss** [comment]
@rtpsw I took one around of review. At the high level I think this makes lot of sense. Left some comments for refinement.

File: cpp/src/arrow/compute/exec/asof_join_node.cc (hunk 5 above)
Comment (rtpsw) [reply cpp/src/arrow/compute/exec/asof_join_node.cc:1040]: Note that the code here already appeared in the pre-PR code, just moved a bit.

I suspect going down the path you describe, which I'm willing to try, would be more complicated than you suggest because even `AsofJoinNodeOptions` does not currently support distinct field-references per input (be them index- or name- based). Therefore, I think the required fix would involve many places in the code (and we would normally have allocated a separate task for such a fix).

File: cpp/proto/substrait/extension_rels.proto (hunk 2 above)
Comment (rtpsw) [reply cpp/proto/substrait/extension_rels.proto:29]: > We are using StructField for this so I don't think we can use name for this, right?

IIRC, Arrow Substrait currently only supports [`StructField` message](https://github.com/substrait-io/substrait/blob/f3f6bdc947e689e800279666ff33f118e42d2146/proto/substrait/algebra.proto#L856) but Substrait itself also provides a [`MapKey` message](https://github.com/substrait-io/substrait/blob/f3f6bdc947e689e800279666ff33f118e42d2146/proto/substrait/algebra.proto#L855) that can be used to reference by name.

> I think it probably makes more sense to have this only be a repeated field in case that on column doesn't share the same column index for each table

If we go down this path (see also [my comment here](https://github.com/apache/arrow/pull/14385#discussion_r994865351)), I think we'd also need to repeat the by-key, i.e., to wrap its existing repetition as a message and repeat that.

File: cpp/src/arrow/compute/exec/asof_join_node.cc (hunk 6 above)
Comment (rtpsw) [reply cpp/src/arrow/compute/exec/asof_join_node.cc:961]: Yes, this happens (and I observed it before adding the code change here) when the by-key is empty.

File: cpp/src/arrow/compute/exec/asof_join_node.cc (hunk 7 above)
Comment (rtpsw) [reply cpp/src/arrow/compute/exec/asof_join_node.cc:1045]: Sorry, which comment? You're referring to `GetIndicesOfByKey` but it's here.

File: cpp/src/arrow/engine/substrait/options.h (hunk 9 above)
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/options.h:39]: It's a [visibility macro](https://github.com/apache/arrow/pull/14385/files#diff-e803a80147d4db13f4a3eee8126ce922ca91a6c6c2295ca3c60d292514c66d1fR22). It tells the linker to export the symbols for the construct following it.

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 12]
```diff
@@ -3187,5 +3198,164 @@ TEST(Substrait, IsthmusPlan) {
...
+                "selection": {
+                  "directReference": {
+                    "structField": {
+                      "field": 0,
+                    }
+                  },
+                  "rootReference": {}
+                }
+              },
+              "by": [
```
[the remark this answers is not in the document]
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/serde_test.cc:3305]: As noted in [this post](https://github.com/apache/arrow/pull/14385#discussion_r994865351), `AsOfJoinNode` code currently does not support distinct per-input references.

[the remark this answers is not in the document]
Comment (icexelloss) [reply cpp/src/arrow/engine/substrait/serde_test.cc:3305]: I am confused - 
https://github.com/apache/arrow/blob/ec579df631deaa8f6186208ed2a4ebec00581dfa/cpp/src/arrow/compute/exec/options.h#L411

Looks like "by" is one per table and "on" is only one for all the tables. So shouldn't substrait message contains multiple entries for "by" key?

[the remark this answers is not in the document]
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/serde_test.cc:3305]: No, both are common to all tables. The "by" field is a vector corresponding to columns, i.e., one can select multiple columns using this vector into the "by" key. If we want per-input field-references, we'd need a vector "on" and a vector-of-vector "by".

[the remark this answers is not in the document]
Comment (icexelloss) [reply cpp/src/arrow/engine/substrait/serde_test.cc:3305]: I see - this makes sense now.

Unfortunately the limitation is too strict I think (requiring all input tables to have the same column index for the by key) 
The restriction for the on-key is OK - we have restrictions for "time" to be the first column. However, we don't have restriction on the "by-key" column - it can have different column index for different input tables.

I think we need to fix this problem in order to Asof Join to be useful - Since we are deciding the shape of the substrait message now - I think it makes sense to fix both message format and asof join options format in this PR.

File: cpp/src/arrow/engine/substrait/options.cc (hunk 8 above)
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/options.cc:64]: As discussed, the vector refers to columns, so it is a single "by" key.

File: cpp/src/arrow/engine/substrait/options.h (hunk 11 above)
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/options.h:74]: Not sure, but I think it's fine to leave as is in this PR.

**Comment by rtpsw** [comment]
CI failures look unrelated the changes here. @icexelloss, this may be ready to go, WDYT?

File: cpp/proto/substrait/extension_rels.proto [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+package arrow.substrait;
+
+import "substrait/algebra.proto";
+
+option csharp_namespace = "Arrow.Substrait";
+option go_package = "github.com/apache/arrow/substrait";
+option java_multiple_files = true;
+option java_package = "io.arrow.substrait";
+
+message AsOfJoinRel {
```
Comment (icexelloss) [inline cpp/proto/substrait/extension_rels.proto:28]: Can you add some documentation for the message?

File: cpp/src/arrow/engine/substrait/expression_internal.cc [hunk 14]
```diff
@@ -52,16 +52,16 @@ Id NormalizeFunctionName(Id id) {
 
 }  // namespace
 
-Status DecodeArg(const substrait::FunctionArgument& arg, uint32_t idx,
+Status DecodeArg(const ::substrait::FunctionArgument& arg, uint32_t idx,
```
Comment (icexelloss) [inline cpp/src/arrow/engine/substrait/expression_internal.cc:55]: Hmm..Are these style change orthogonal?

File: cpp/proto/substrait/extension_rels.proto [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+  repeated AsOfJoinKeys input_keys = 1;
```
Comment (icexelloss) [inline cpp/proto/substrait/extension_rels.proto:29]: I think `keys` is a little concise/betters here.

**Comment by icexelloss** [comment]
@rtpsw Looks pretty good! There seems to be many style changes that seems orthogonal. Are those changes automatically made? Is it easy to separate out the style change from the actual change?

File: cpp/src/arrow/engine/substrait/expression_internal.cc (hunk 14 above)
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/expression_internal.cc:55]: Unfortunately, no. I observed many CI jobs failures without these changes.

**Comment by rtpsw** [comment]
> > @rtpsw Looks pretty good! There seems to be many style changes that seems orthogonal. Are those changes automatically made? Is it easy to separate out the style change from the actual change?
> 
> If we insist, we could take these style changes out to a separate PR, get that merged, and then come back here to rebase.

Anja suggested changing `arrow::engine::substrait` to `arrow::engine::substrait_ext` to avoid MSVC name conflicts with symbols in `::substrait`. This should allow us to keep using `substrait::` instead of `::substrait::` and avoid a large orthogonal change.

**Comment by icexelloss** [comment]
> > > @rtpsw Looks pretty good! There seems to be many style changes that seems orthogonal. Are those changes automatically made? Is it easy to separate out the style change from the actual change?
> > 
> > 
> > If we insist, we could take these style changes out to a separate PR, get that merged, and then come back here to rebase.
> 
> Anja suggested changing `arrow::engine::substrait` to `arrow::engine::substrait_ext` to avoid MSVC name conflicts with symbols in `::substrait`. This should allow us to keep using `substrait::` instead of `::substrait::` and avoid a large orthogonal change.

Does `arrow::engine::substrait::extension` work?

**Comment by rtpsw** [comment]
> Does `arrow::engine::substrait::extension` work?

My guess is this would cause MSVC just the same because when it sees a `substrait::*` symbol it would look under `arrow::engine::substrait` and won't fall back to `::substrait`.

**Comment by rtpsw** [comment]
> @rtpsw Looks pretty good! There seems to be many style changes that seems orthogonal. Are those changes automatically made? Is it easy to separate out the style change from the actual change?

@icexelloss, I managed to minimize the orthogonal change (`namespace substrait = ::substrait` added to a few files). I think this should be enough to make reviewing convenient.

File: cpp/proto/substrait/extension_rels.proto (hunk 13 above)
Comment (rtpsw) [reply cpp/proto/substrait/extension_rels.proto:29]: Done.

File: cpp/proto/substrait/extension_rels.proto (hunk 15 above)
Comment (rtpsw) [reply cpp/proto/substrait/extension_rels.proto]: Done.

**Comment by rtpsw** [comment]
The commit history got messed up. I'll try to open a fresh PR.

**Comment by icexelloss** [comment]
@rtpsw You don't need to open a new PR - you can just force push to your branch ARROW-17980 and fix the commit history
